src/cargo/core/compiler/mod.rs

```diff
@@ -694,6 +694,11 @@ fn prepare_rustc(build_runner: &BuildRunner<'_, '_>, unit: &Unit) -> CargoResult
     base.inherit_jobserver(&build_runner.jobserver);
     build_deps_args(&mut base, build_runner, unit)?;
     add_cap_lints(build_runner.bcx, unit, &mut base);
+    if build_runner.bcx.gctx.nightly_features_allowed {
```

Comment:
> Unsure if this affects people's workflow, so this behavior is only
enabled on nightly for collectin feedback. If everything goes well,
the nightly gate will be removed after a few of releases.

When we discussed this in the Cargo team meeting, one idea was a short-term env variable to opt-in to this and then we transition to opt-out from this.

What were your thoughts on this direction?

Reply:
Should we go with what we've done to git2-to-gitx transition for cargo-publish, like having an opt-out env var for a while? I slightly doubt that people will use a short-live opt-in environment variable, not to say it's pretty hard to get people onboard and solicit feedback.

Reply:
> Should we go with what we've done to git2-to-gitx transition for cargo-publish, like having an opt-out env var for a while? I slightly doubt that people will use a short-live opt-in environment variable, not to say it's pretty hard to get people onboard and solicit feedback.

I can understand skipping opt-in.

If you are referring to doing that on stable, the one difference is that git2/gix were meant to be behavior parity while in this case we are changing behavior.  Having an opt-out on stable buys people time to update if it breaks them but it doesn't let us take feedback and change it as it will already be stable.

If you are only referring to nightly, then I'm fine.

Reply:
Nightly only for now. Let's see how it goes.

> Having an opt-out on stable buys people time to update if it breaks them but it doesn't let us take feedback and change it as it will already be stable.

To me this is the way to go. We have done similar process for config array merging, and expected it to have a larger impact. Arlo ever wrote an [inside rust post](https://blog.rust-lang.org/inside-rust/2023/08/24/cargo-config-merging.html) for it. However, we have not heard back from anyone, not even a linked issue to [the original PR](https://github.com/rust-lang/cargo/pull/12515). Yet, no need to decide how to stabilize now.

Reply:
An environment variable is added `__CARGO_RUSTC_ORIG_ARGS_PRIO`.